### PR TITLE
Add ResourceInfo for allocating singletons

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Add a struct called `ResourceInfo` that will hold all information regarding placing singletons and array components on the given resources. This information includes whether to avoid placing singletons/array components on the global 0th processor, what processors to place the singletons on, and whether they are exclusively on that processor using a `SingletonPack` (see #3962). 

This struct uses a custom algorithm to spread the singletons in an executable evenly over the available resources to balance communication and computational costs. Also, this struct will gather info on which singletons want to be exclusively on a processor and return a set of processors to ignore that can be given to the `procs_to_ignore` argument of the `allocate_array` function of an array component (see #3950).

See the documentation of `ResourceInfo` for a description of how the algorithm places singletons.

I intend for this struct to be stored as a member variable of Main and be used to allocate the singletons for all executables. That change will happen in a future PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by knelli2** [comment]
Because I intend to put this in Main, I'm requesting a review from all the active core-devs as I want their input.

I also added the "new design" label because the algorithm I created for placing the singletons is somewhat complicated and I'd like feedback on it. If you have too many questions/comments or ones that may be too long to type out, I'm happy to set up a meeting/call and talk about my design decisions.

File: src/Parallel/ResourceInfo.hpp [hunk 1]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+  // For each singleton (whether it has a SingletonInfo or not), maps whether
+  // it's exclusive and what proc it is on.
+  tuples::tagged_tuple_from_typelist<local_tags> singleton_map_{};
+};
+
+template <typename Metavariables>
+void ResourceInfo<Metavariables>::build_singleton_map(
+    const Parallel::GlobalCache<Metavariables>& cache) {
+  const size_t num_procs =
+      static_cast<size_t>(Parallel::number_of_procs(cache));
```
Comment (knelli2) [inline src/Parallel/ResourceInfo.hpp]: Once a decision on #3976 is made, I'll update this PR to use those functions as I do a lot of `static_cast`ing here. Until then, I'll just leave these as is.

File: src/Parallel/ResourceInfo.hpp [hunk 2]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+
+template <typename Component>
+using component_has_singleton_info_tag =
+    tmpl::list_contains<init_tags<Component>,
+                        Parallel::Tags::SingletonInfo<Component>>;
+
+template <typename Metavariables>
+using singleton_components =
+    tmpl::filter<typename Metavariables::component_list,
+                 tmpl::bind<Parallel::is_singleton, tmpl::_1>>;
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: No `bind`.

File: src/Parallel/ResourceInfo.hpp [hunk 3]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+
+template <typename InitTags>
+using list_contains_avoid_global_proc_0 =
+    tmpl::list_contains<InitTags, Parallel::Tags::AvoidGlobalProc0>;
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: Consider inlining some of these really simple metafuntions that are only used once.

File: src/Parallel/ResourceInfo.hpp [hunk 4]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+    p | singleton_map_;
+  }
+
+  /// Returns whether we should avoid placing array elements and singletons on
+  /// the global zeroth proc. Default `false`.
+  bool avoid_global_proc_0() const { return avoid_global_proc_0_; }
+
+  /// Return a SingletonInfoHolder corresponding to `Component`
+  template <typename Component>
+  auto get_singleton_info() {
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: const?

File: src/Parallel/ResourceInfo.hpp [hunk 5]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+    const auto& singleton_map =
+        tuples::get<LocalTag<Component>>(singleton_map_);
+    return SingletonInfoHolder<Component>{
+        {static_cast<int>(*singleton_map.second)}, singleton_map.first};
+  }
+
+  /// Returns a `std::unordered_set<size_t>` of processors that array components
+  /// should avoid placing elements on. This should be passed to the
+  /// `allocate_array` function of the array component
+  std::unordered_set<size_t> procs_to_ignore() {
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: const?  Return reference?

File: src/Parallel/ResourceInfo.hpp [hunk 6]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+    if (not singleton_map_has_been_set_) {
+      singleton_map_not_built();
+    }
+    return procs_to_ignore_;
+  }
+
+  /// Returns the proc that the singleton `Component` should be placed on.
+  template <typename Component>
+  size_t proc_for() {
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: const?

File: src/Parallel/ResourceInfo.hpp [hunk 7]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+
+  // Check if any singletons that requested to be on specific proc requested to
+  // be on a proc beyond the last proc.
+  tmpl::for_each<singletons>([this, &num_procs](const auto component_v) {
+    using component = tmpl::type_from<decltype(component_v)>;
+    auto& singleton_map = tuples::get<LocalTag<component>>(singleton_map_);
+    const auto proc = singleton_map.second;
+
+    if (proc.has_value() and *proc > num_procs - 1) {
+      ERROR("Singleton %s requested to be placed on proc "
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: Fix `%s`

File: src/Parallel/ResourceInfo.hpp [hunk 8]
```diff
@@ -186,4 +205,698 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ *    number of nodes. We say "as evenly as possible" because this depends on
+ *    the requested exclusive singletons. For example, if we have 4 nodes and 5
+ *    cores per node, the number of requested exclusive singletons on each node
+ *    is (0, 1, 4, 1), and we have 3 auto exclusive singletons to place, the
+ *    best distribution of exclusive singletons we can achieve given our
+ *    constraints is (2, 2, 4, 1). Clearly this is not the *most* evenly
+ *    distributed the exclusive singletons could be. However, this *is* the most
+ *    evenly distributed they could be given the starting distribution from the
+ *    input file.
+ *    3. Allocate auto nonexclusive singletons, distributing the total number of
```
Comment (wthrowe) [inline src/Parallel/ResourceInfo.hpp]: This isn't formatting correctly, not sure why.  (It is coming out as 2a instead of 3.)

File: src/Parallel/Tags/ResourceInfo.hpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  static std::string name() { return pretty_type::name<ParallelComponent>(); }
+
+  static constexpr bool pass_metavariables = true;
+
+  template <typename Metavariables>
+  using option_tags =
+      tmpl::list<Parallel::OptionTags::ResourceInfo<Metavariables>>;
+
+  // Can't take a const& here because the get_singleton_info() member isn't
+  // marked const
```
Comment (wthrowe) [inline src/Parallel/Tags/ResourceInfo.hpp]: Why not?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Parallel/ResourceInfo.hpp (hunk 8 above)
Comment (knelli2) [reply src/Parallel/ResourceInfo.hpp]: Strange...it looked correct for me locally. I fixed the indent, hopefully that helps.

Comment (knelli2) [reply src/Parallel/ResourceInfo.hpp]: Ok yeah that looked like it fixed it.

Comment (wthrowe) [reply src/Parallel/ResourceInfo.hpp]: It may be version-dependent or something.  I occasionally find that my local version and the CI generated docs disagree.

**Review by wthrowe — DISMISSED** [review]
Squash.

**Review by wthrowe — DISMISSED** [review]
As noted above, can be cleaned slightly with #3976 merged, but neither PR actually depends on or conflicts with the other.

File: src/Parallel/ResourceInfo.hpp [hunk 10]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ * \brief Holds resource info for all singletons and for avoiding placing array
+ * elements/singletons on the global proc 0.
+ *
+ * \details This can be used for placing all singletons in an executable. To
+ * have a singleton specify resource information, add the
+ * `Parallel::Tags::SingletonInfo` tag to the `initialization_tags` type alias
+ * in the singleton. To specify whether to avoid placing array elements and
+ * singletons on the global proc 0, add the `Parallel::Tags::AvoidGlobalProc0`
+ * tag to the `initialization_tags` of any parallel component in your
+ * executable.
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp:255]: Wouldn't this only work for singletons and arrays? I think by definition groups and nodegroups have something on all cores

File: src/Parallel/ResourceInfo.hpp [hunk 11]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ * Several consistency checks are done during option parsing to avoid user
+ * error. However, some checks can't be done during option parsing because the
+ * number of nodes/procs is needed to determine if there is an inconsistency.
+ * These checks are done during runtime, just before the map of singletons is
+ * created.
+ *
+ * To automatically place singletons, we use a custom algorithm that will
+ * distribute singletons evenly over the number of nodes, and evenly over the
+ * procs on a node. This will help keep communication costs down by distributing
+ * the workload over all of the communication cores (one comm core per charm
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `comm` -> `communication` (just to make it easier to read for folks less familiar with Charm++ names)

File: src/Parallel/ResourceInfo.hpp [hunk 12]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ * node), and ensure that our resources are being maximally utilized (i.e. one
+ * core doesn't have all the singleton on it).
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `singleton` -> `singletons` (pluralize)

File: src/Parallel/ResourceInfo.hpp [hunk 13]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ *
+ * Defining some terminology for singletons: `requested` means that a specific
+ * processor was requested in the input file; `auto` means that the processor
+ * should be chosen automatically by the algorithm; `exclusive` means that no
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: I'm a bit worried that `by the algorithm` will be confused with the `Algorithm` class. I'm not sure I have a request here, but something to think about. Maybe just removing `by the algorithm` :shrug:

File: src/Parallel/ResourceInfo.hpp [hunk 14]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ * other singletons or array elements should be placed on this singletons
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `singletons` -> `singleton's` (possessive)?

File: src/Parallel/ResourceInfo.hpp [hunk 15]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ * processor; `nonexclusive` means that you *can* place other singletons or
+ * array elements on this singletons processor. The algorithm is as follows:
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `singletons` -> `singleton's` (possessive)

Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `algorithm` -> `algorithm that distributes the singletons`

File: src/Parallel/ResourceInfo.hpp [hunk 16]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+ *
+ * 1. Allocate all singletons that requested specific processors, both exclusive
+ *    and nonexclusive. This is done during option parsing.
+ * 2. Allocate auto exclusive singletons, distributing the total number of
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: maybe ticks around `auto`?

File: src/Parallel/ResourceInfo.hpp [hunk 17]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+        "0."};
+  };
+
+  using options = tmpl::flatten<tmpl::list<
+      tmpl::conditional_t<detail::avoid_global_proc_0<Metavariables>,
+                          tmpl::list<AvoidGlobalProc0>, tmpl::list<>>,
+      tmpl::conditional_t<tmpl::size<singletons_with_info>::value != 0,
+                          tmpl::list<Singletons>, tmpl::list<>>>>;
+  static constexpr Options::String help = {
+      "Resource options for a simulation."};
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: While not wrong, something a bit more descriptive would probably be helpful :)

File: src/Parallel/ResourceInfo.hpp [hunk 18]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+
+  /// The main constructor. All other constructors that take options will call
+  /// this one. This constructor holds all checks able to be done during option
+  /// parsing.
+  ResourceInfo(const bool avoid_global_proc_0,
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp:387]: Can you move the all the constructors and member functions out of the class definition? As-is they are implicitly marked `inline`, which can add compilation overheads, especially for such long, complicated constructors

File: src/Parallel/ResourceInfo.hpp [hunk 19]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+    p | num_requested_exclusive_singletons_;
+    p | num_requested_nonexclusive_singletons_;
+    p | requested_nonexclusive_procs_;
+    p | procs_to_ignore_;
+    p | singleton_map_;
+  }
+
+  /// Returns whether we should avoid placing array elements and singletons on
+  /// the global zeroth proc. Default `false`.
+  bool avoid_global_proc_0() const { return avoid_global_proc_0_; }
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp:415]: this function can stay inlined

File: src/Parallel/ResourceInfo.hpp [hunk 20]
```diff
@@ -186,4 +205,687 @@ struct SingletonPack<tmpl::list<ParallelComponents...>> {
...
+          std::find_if(first_proc_iter, first_proc_next_node_iter,
+                       [&min_num_singletons_on_a_proc](const auto& proc_opt) {
+                         return proc_opt.has_value() and
+                                *proc_opt == min_num_singletons_on_a_proc;
+                       });
+
+      // Get the index of the overall vector. We need this because we're going
+      // to be indexing two separate vectors, otherwise we could have just used
+      // the value of the iterator
+      size_t proc = static_cast<size_t>(std::distance(
```
Comment (nilsdeppe) [inline src/Parallel/ResourceInfo.hpp]: `const size_t`?

File: src/Parallel/Tags/ResourceInfo.hpp [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+#include "Utilities/TMPL.hpp"
+
+namespace Parallel {
+namespace OptionTags {
+/// \ingroup ParallelGroup
+/// Options group for resource allocation
+template <typename Metavariables>
+struct ResourceInfo {
+  using type = Parallel::ResourceInfo<Metavariables>;
+  static constexpr Options::String help = {"Options for allocating resources."};
```
Comment (nilsdeppe) [inline src/Parallel/Tags/ResourceInfo.hpp]: Agreed, but could this be more specific?

File: src/Parallel/Tags/ResourceInfo.hpp [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+};
+}  // namespace OptionTags
+
+namespace Tags {
+/// \ingroup DataBoxTagsGroup
```
Comment (nilsdeppe) [inline src/Parallel/Tags/ResourceInfo.hpp]: I personally think the docs is cleaner not sticking everything into the groups, so my suggestion is to not put these tags into the groups. We definitely haven't been consistent, but as the number of tags has increased, putting them all in the Doxygen group I've found to make the doxygen group very clutter :shrug:

File: src/Parallel/Tags/ResourceInfo.hpp [hunk 23]
```diff
@@ -0,0 +1,65 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <string>
+
+#include "Options/Options.hpp"
```
Comment (nilsdeppe) [inline src/Parallel/Tags/ResourceInfo.hpp:9]: You need the include for `db::SimpleTag`

File: tests/Unit/Parallel/Test_ResourceInfo.cpp [hunk 24]
```diff
@@ -119,16 +182,793 @@ void test_singleton_pack() {
...
+  for (size_t i = 0; i < singletons.size(); i++) {
+    const bool exclusive = singletons[i].first;
+    const int proc = singletons[i].second;
+    option_str += "  FakeSingleton" + get_output(i) + ":\n";
+    option_str +=
+        "    Proc: " + (proc == -1 ? "Auto"s : get_output(proc)) + "\n";
+    option_str += "    Exclusive: " + (exclusive ? "true"s : "false"s) + "\n";
+  }
+
+  return TestHelpers::test_option_tag<OptionTags::ResourceInfo<Metavariables>>(
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_ResourceInfo.cpp]: add a `serialize_and_deserialize` here so you can simultaneously test all the serialization?

File: tests/Unit/Parallel/Test_ResourceInfo.cpp [hunk 25]
```diff
@@ -119,16 +182,793 @@ void test_singleton_pack() {
...
+      option_str);
+}
+
+constexpr size_t num_singletons = 7;
+using metavars = MetavariablesBoth<0, 1, 2, 3, 4, 5, 6>;
+template <size_t Index>
+using component = FakeSingleton<metavars, Index>;
+
+void check_resource_info(
+    Parallel::GlobalCache<metavars>& cache, bool avoid_global_proc_0,
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_ResourceInfo.cpp]: `const &`?

File: tests/Unit/Parallel/Test_ResourceInfo.cpp [hunk 26]
```diff
@@ -119,16 +182,793 @@ void test_singleton_pack() {
...
+              expected_singletons[index].first);
+        CHECK(static_cast<int>(singleton_info.proc().value()) ==
+              expected_singletons[index].second);
+        CHECK(resource_info.template proc_for<component<index>>() ==
+              static_cast<size_t>(expected_singletons[index].second));
+      });
+}
+
+template <typename Gen>
+void test_single_node_multi_core(gsl::not_null<Gen*> gen) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_ResourceInfo.cpp]: `const gsl`

File: tests/Unit/Parallel/Test_ResourceInfo.cpp [hunk 27]
```diff
@@ -119,16 +182,793 @@ void test_singleton_pack() {
...
+    const std::vector<std::pair<bool, int>> expected{
+        {false, 2}, {false, 2}, {false, 1}, {false, 1},
+        {true, 0},  {false, 1}, {false, 2}};
+
+    check_resource_info(cache, false, singletons, expected);
+  }
+}
+
+template <typename Gen>
+void test_multi_node_multi_core(gsl::not_null<Gen*> gen) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_ResourceInfo.cpp]: `const gsl`

File: tests/Unit/Parallel/Test_ResourceInfo.cpp [hunk 28]
```diff
@@ -119,16 +182,793 @@ void test_singleton_pack() {
...
+    std::vector<std::pair<bool, int>> expected{
+        {false, 4}, {true, 2}, {false, 4}, {false, 4},
+        {true, 3},  {true, 0}, {true, 1}};
+
+    check_resource_info(local_cache, false, singletons, expected);
+  }
+}
+
+template <typename Gen>
+void test_multi_node_multi_core_large(gsl::not_null<Gen*> gen) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_ResourceInfo.cpp]: `const gsl`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Parallel/ResourceInfo.hpp (hunk 10 above)
Comment (knelli2) [reply src/Parallel/ResourceInfo.hpp:255]: At the beginning of the sentence I say `To specify whether to avoid placing array elements and singletons on the global proc 0, ...` so I'm not indicating that (node)groups won't be placed on global proc 0. I just check the initialization tags of all components for the AvoidGlobalProc0 tag, but the tag will only have an effect on the placement of singletons and arrays.

Should I reword this to make it clearer?

Comment (nilsdeppe) [reply src/Parallel/ResourceInfo.hpp:255]: Hmm, agreed. Maybe `of the... components` (last pluralize). But up to you, I agree you do give the restriction at the beginning so if someone encounters and issue they can reread and it should clarify :)

File: src/Parallel/Tags/ResourceInfo.hpp (hunk 22 above)
Comment (knelli2) [reply src/Parallel/Tags/ResourceInfo.hpp]: I removed them from the DataBoxTags group, but left them in the Parallel group because there are less things in the Parallel group and because I think they are important enough tags that they should be discoverable with the rest of the parallel things.

**Comment by nilsdeppe** [comment]
LGTM, please squash

**Comment by knelli2** [comment]
Rebased and cleaned up parts dealing with #3976. No other changes

**Comment by knelli2** [comment]
@kidder Did you want to review this?

**Comment by kidder** [comment]
I have not done a detailed review, but am happy with the design.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@wthrowe if you're happy with this could you merge it? I just want to make sure none of my requested changes are anything you disagree with :)

**Review by wthrowe — APPROVED** [review]
